### PR TITLE
New client lib (Review ONLY)

### PR DESCRIPTION
**Pull request by dara123**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -19,9 +19,21 @@ and then use according to documentation.
 ```php
 require('./client/php/rebilly_api.php');
 
-$customer = new Rebilly\v2_1\Customer();
-$customer->setEnvironment(RebillyRequest::ENV_SANDBOX);
-$customer->setApiKey(API_SECRET_KEY);
+// create new customer
+\Rebilly\Api\Request::factory([
```
Comment (tonydspaniard) [inline README.md]: not sure we need to add Api as a namespace. IMHO I would add root objects in Rebilly namespace only. No need for `Api`. This is the API library.

File: README.md [hunk 2]
```diff
@@ -19,9 +19,21 @@ and then use according to documentation.
...
+]);
+
+$model = new \Rebilly\Model\Customer([
+	'id' => 'custA5',
+	'email' => 'custA5@email.com',
+	'firstName' => 'johnABC',
+	'lastName' => 'DoeABC',
+]);
+
+$customer = \Rebilly\Api\Customer::create($model);
```
Comment (tonydspaniard) [inline README.md]: I think this approach is confusing. First, the static factory method of `Request::factory` is not returning an instance, which is weird, most of the factory methods where it requires some configuration do return an instance of a class. Secondly, the fact that we use `Customer::create` to decorate a request I do not think is better than do something like (just brainstorming): `$requestInstance->customers->create($model)`

Comment (dara123) [inline README.md]: I like this idea. :+1:

Comment (tonydspaniard) [inline README.md]: `$rq->customers` is a decorator method accessed via magic `Request::__get()` that knows what is the API endpoint to commit the request. Just an idea.

File: lib/Api/Customer.php [hunk 3]
```diff
@@ -0,0 +1,71 @@
+<?php
+namespace Rebilly\Api;
+
+use Exception;
+use Rebilly\Model\Customer as CustomerModel;
+
+class Customer extends Request
```
Comment (tonydspaniard) [inline lib/Api/Customer.php]: IMHO I believe that this class should not be a child of `Request` but a child of an `EndPoint` class (or something like that) where it contains the methods that will help the request to determine where to execute the request and (maybe) returning the appropriate Response -ie CustomerCreateResponse.

File: lib/Api/Customer.php [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+     * @return CustomerModel
+     */
+    public static function create(CustomerModel $customer)
+    {
+        if (!empty($customer->id)) {
+            self::setApiEndPoint(self::END_POINT . $customer->id);
+            $response = self::sendPutRequest($customer->getDataObject());
+        } else {
+            self::setApiEndPoint(self::END_POINT);
+            $response = self::sendPostRequest($customer->getDataObject());
```
Comment (tonydspaniard) [inline lib/Api/Customer.php]: I think this should be executed in the `Request object` not here.

File: lib/Api/Request.php [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        if (self::$httpClient === null) {
+            self::$httpClient = new HttpClient($config);
+        }
+
+        return self::$httpClient;
+    }
+
+    public static function sendGetRequest()
+    {
+        return self::getCurlAdaptor()->sendRequest(self::getHttpClient(), HttpClient::METHOD_GET);
```
Comment (tonydspaniard) [inline lib/Api/Request.php]: On many of the current famous `HTTP Request | Response` libraries (like Guzzle) the HttpClient is actually the one that makes the requests. So, is kind of confusing that we have a `Request` class that needs an `HttpClient` which only provides configuration information.

File: lib/Api/Request.php [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    public static function sendPostRequest($request)
+    {
+        return self::getCurlAdaptor()->sendRequest(self::getHttpClient(), HttpClient::METHOD_POST, $request);
```
Comment (tonydspaniard) [inline lib/Api/Request.php]: it makes more sense (if we using our own client) that the `Request` is actually a child class of the adapter and it has the decorator methods that the selected adapter does not provide -ie `sendPostRequest` on this case.

File: lib/Model/Base.php [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+abstract class Base
+{
+    private $statusCode;
+    private $errorMessage;
+
+    /**
+     * Set model object value
+     * @param array $json
+     */
+    abstract public function setDataObject($json);
```
Comment (tonydspaniard) [inline lib/Model/Base.php]: I propose: `static buildFromArray()` or `static buildFromJsonArray()` or `setAttributes`

File: lib/Model/Base.php [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+    /**
+     * Get model object value
+     * @return mixed
+     */
+    abstract public function getDataObject();
```
Comment (tonydspaniard) [inline lib/Model/Base.php]: I propose extend from `Arrayable` Interface and use `toArray()`

File: README.md (hunk 1 above)
Comment (2e3s) [inline README.md]: I would except from a factory method some result, some instance of class.

Comment (tonydspaniard) [inline README.md]: @2e3s  exactly

File: examples/Customer.php [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+
+class Customer
+{
+    /**
+     * Get customer
+     */
+    public function getCustomer()
+    {
+        // Set ApiKey and environment
+        $client = \Rebilly\Rebilly::factory([
```
Comment (tonydspaniard) [inline examples/Customer.php]: Why don't you just have a private method returning the instance :) ?

File: lib/Api/Blacklist.php [hunk 10]
```diff
@@ -0,0 +1,56 @@
+<?php
+namespace Rebilly\Api;
+
+use Rebilly\Model\Blacklist as BlacklistModel;
+
+class Blacklist extends Request
+{
+    /** public const */
+    const TYPE_EMAIL = 'email';
```
Comment (tonydspaniard) [inline lib/Api/Blacklist.php]: Don't add constants here. I think is better to add it to `Rebilly\Model\Blacklist\Type` class. Sorry, maybe too much DDD and SRP.

File: lib/Api/Customer.php [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+    const END_POINT = 'customers/';
+
+    /**
+     * Create new customer
+     * @return CustomerModel
+     */
+    public function create(CustomerModel $customer)
+    {
+        if (!empty($customer->id)) {
+            $this->setApiEndPoint(self::END_POINT . $customer->id);
```
Comment (tonydspaniard) [inline lib/Api/Customer.php]: Why we do constantly need to set the Api End Point? There is something wrong here, the API end point should be set on the constructor, where its parameters attached to it as needed -ie `$customer->id`

File: lib/Api/Customer.php (hunk 3 above)
Comment (adamaltman) [inline lib/Api/Customer.php]: Why a child of anything?

**Pull request closed without merging**